# This patch includes an assortment of various commits from PortAudio's master
# branch, particularly to enable WASAPI loopback support for PortAudio 19.7. It
# is temporary and not meant to be upstreamed.
diff --git a/build/msvc/portaudio.def b/build/msvc/portaudio.def
index 6ecb142..39d425c 100644
--- a/build/msvc/portaudio.def
+++ b/build/msvc/portaudio.def
@@ -56,3 +56,8 @@ PaWasapi_SetStreamStateHandler      @68
 PaWasapiWinrt_SetDefaultDeviceId    @67
 PaWasapiWinrt_PopulateDeviceList    @69
 PaWasapi_GetIMMDevice               @70
+PaWasapi_IsLoopback                 @71
+PaWinMME_GetStreamInputHandleCount  @72
+PaWinMME_GetStreamInputHandle       @73
+PaWinMME_GetStreamOutputHandleCount @74
+PaWinMME_GetStreamOutputHandle      @75
diff --git a/cmake_support/template_portaudio.def b/cmake_support/template_portaudio.def
index 9cf0dc3..0e6cb8b 100644
--- a/cmake_support/template_portaudio.def
+++ b/cmake_support/template_portaudio.def
@@ -59,3 +59,8 @@ PaUtil_SetDebugPrintFunction        @55
 @DEF_EXCLUDE_WASAPI_SYMBOLS@PaWasapiWinrt_SetDefaultDeviceId    @67
 @DEF_EXCLUDE_WASAPI_SYMBOLS@PaWasapiWinrt_PopulateDeviceList    @69
 @DEF_EXCLUDE_WASAPI_SYMBOLS@PaWasapi_GetIMMDevice               @70
+@DEF_EXCLUDE_WASAPI_SYMBOLS@PaWasapi_IsLoopback                 @71
+@DEF_EXCLUDE_WMME_SYMBOLS@PaWinMME_GetStreamInputHandleCount    @72
+@DEF_EXCLUDE_WMME_SYMBOLS@PaWinMME_GetStreamInputHandle         @73
+@DEF_EXCLUDE_WMME_SYMBOLS@PaWinMME_GetStreamOutputHandleCount   @74
+@DEF_EXCLUDE_WMME_SYMBOLS@PaWinMME_GetStreamOutputHandle        @75
diff --git a/include/pa_win_wasapi.h b/include/pa_win_wasapi.h
index c046afd..4f97917 100644
--- a/include/pa_win_wasapi.h
+++ b/include/pa_win_wasapi.h
@@ -84,7 +84,10 @@ typedef enum PaWasapiFlags
        playback formats that do not match the current configured system settings.
        this is in particular required for streams not matching the system mixer sample rate.
        only applies in Shared mode. */
-    paWinWasapiAutoConvert              = (1 << 6)
+    paWinWasapiAutoConvert              = (1 << 6),
+
+    /* use Passthrough mode for sending encoded audio data in PCM containers to the audio device */
+    paWinWasapiPassthrough              = (1 << 7),
 }
 PaWasapiFlags;
 #define paWinWasapiExclusive             (paWinWasapiExclusive)
@@ -94,6 +97,7 @@ PaWasapiFlags;
 #define paWinWasapiThreadPriority        (paWinWasapiThreadPriority)
 #define paWinWasapiExplicitSampleFormat  (paWinWasapiExplicitSampleFormat)
 #define paWinWasapiAutoConvert           (paWinWasapiAutoConvert)
+#define paWinWasapiPassthrough           (paWinWasapiPassthrough)
 
 
 /* Stream state.
@@ -304,6 +308,59 @@ typedef enum PaWasapiStreamOption
 PaWasapiStreamOption;
 
 
+/** Passthrough format.
+
+    Format ids are taken from the Microsoft documentation "Representing Formats for IEC 61937 Transmissions"
+    and are composed by such formula where GUID is the guid of passthrough format:
+    GUID.Data1 << 16 | GUID.Data2.
+
+ @version Available as of 19.8.0
+*/
+typedef enum PaWasapiPassthroughFormat
+{
+    ePassthroughFormatPcmIec60958           = 0x00000000,
+    ePassthroughFormatDolbyDigital          = 0x00920000,
+    ePassthroughFormatMpeg1                 = 0x00030cea,
+    ePassthroughFormatMpeg3                 = 0x00040cea,
+    ePassthroughFormatMpeg2                 = 0x00050cea,
+    ePassthroughFormatAac                   = 0x00060cea,
+    ePassthroughFormatDts                   = 0x00080cea,
+    ePassthroughFormatDolbyDigitalPlus      = 0x000a0cea,
+    ePassthroughFormatDolbyDigitalPlusAtmos = 0x010a0cea,
+    ePassthroughFormatDtsHd                 = 0x000b0cea,
+    ePassthroughFormatDtsXE1                = 0x010b0cea,
+    ePassthroughFormatDtsXE2                = 0x030b0cea,
+    ePassthroughFormatDolbyMlp              = 0x000c0cea,
+    ePassthroughFormatDolbyMat20            = 0x010c0cea,
+    ePassthroughFormatDolbyMat21            = 0x030c0cea,
+    ePassthroughFormatWmaPro                = 0x01640000,
+    ePassthroughFormatAtrac                 = 0x00080cea,
+    ePassthroughFormatOneBitAudio           = 0x00090cea,
+    ePassthroughFormatDst                   = 0x000d0cea,
+}
+PaWasapiPassthroughFormat;
+
+
+/** Passthrough details.
+
+    Passthrough details provide direct link to the additional members in WAVEFORMATEXTENSIBLE_IEC61937.
+    Passthrough mode allows to pass encoded data inside the PCM containers to the audio device.
+
+    Detailed description about supported formats and examples are provided in Microsoft documentation:
+    "Representing Formats for IEC 61937 Transmissions".
+
+ @version Available as of 19.8.0
+*/
+typedef struct PaWasapiStreamPassthrough
+{
+    PaWasapiPassthroughFormat formatId;
+    unsigned int encodedSamplesPerSec;
+    unsigned int encodedChannelCount;
+    unsigned int averageBytesPerSec;
+}
+PaWasapiStreamPassthrough;
+
+
 /* Stream descriptor. */
 typedef struct PaWasapiStreamInfo
 {
@@ -349,6 +406,13 @@ typedef struct PaWasapiStreamInfo
      @version Available as of 19.6.0
     */
     PaWasapiStreamOption streamOption;
+
+    /** Passthrough details.
+     @note paWinWasapiPassthrough flag must be specified in PaWasapiStreamInfo::flags to enable Passthrough mode.
+     @see paWinWasapiPassthrough
+     @version Available as of 19.7.0
+    */
+    PaWasapiStreamPassthrough passthrough;
 }
 PaWasapiStreamInfo;
 
@@ -437,7 +501,7 @@ int PaWasapi_GetDeviceMixFormat( void *pFormat, unsigned int formatSize, PaDevic
 int/*PaWasapiDeviceRole*/ PaWasapi_GetDeviceRole( PaDeviceIndex device );
 
 
-/** Get device IMMDevice pointer
+/** Get device IMMDevice pointer.
 
  @param device Device index.
  @param pAudioClient Pointer to pointer of IMMDevice.
@@ -447,6 +511,20 @@ int/*PaWasapiDeviceRole*/ PaWasapi_GetDeviceRole( PaDeviceIndex device );
 PaError PaWasapi_GetIMMDevice( PaDeviceIndex device, void **pIMMDevice );
 
 
+/** Get device loopback state:
+
+    0 - Not loopback,
+    1 - Loopback,
+    negative - PaErrorCode.
+
+ @param device Device index.
+
+ @return Non-negative value indicating loopback state or, a PaErrorCode (which is always negative)
+         if PortAudio is not initialized or an error is encountered.
+*/
+int PaWasapi_IsLoopback( PaDeviceIndex device );
+
+
 /** Boost thread priority of calling thread (MMCSS).
 
     Use it for Blocking Interface only inside the thread which makes calls to Pa_WriteStream/Pa_ReadStream.
diff --git a/src/hostapi/jack/pa_jack.c b/src/hostapi/jack/pa_jack.c
index 124c0f8..b8fe4d4 100644
--- a/src/hostapi/jack/pa_jack.c
+++ b/src/hostapi/jack/pa_jack.c
@@ -1823,4 +1823,3 @@ PaError PaJack_GetClientName(const char** clientName)
 error:
     return result;
 }
-
diff --git a/src/hostapi/wasapi/pa_win_wasapi.c b/src/hostapi/wasapi/pa_win_wasapi.c
index c76f302..a5500d3 100644
--- a/src/hostapi/wasapi/pa_win_wasapi.c
+++ b/src/hostapi/wasapi/pa_win_wasapi.c
@@ -1,7 +1,7 @@
 /*
  * Portable Audio I/O Library WASAPI implementation
  * Copyright (c) 2006-2010 David Viens
- * Copyright (c) 2010-2019 Dmitry Kostjuchenko
+ * Copyright (c) 2010-2023 Dmitry Kostjuchenko
  *
  * Based on the Open Source API proposed by Ross Bencina
  * Copyright (c) 1999-2019 Ross Bencina, Phil Burk
@@ -44,6 +44,7 @@
 */
 
 #include <windows.h>
+#include <stddef.h>
 #include <stdio.h>
 #include <process.h>
 #include <assert.h>
@@ -279,9 +280,26 @@ PA_DEFINE_IID(IPart,                AE2DE0E4, 5BCA, 4F2D, aa, 46, 5d, 13, f8, fd
 PA_DEFINE_IID(IKsJackDescription,   4509F757, 2D46, 4637, 8e, 62, ce, 7d, b9, 44, f5, 7b);
 
 // Media formats:
-__DEFINE_GUID(pa_KSDATAFORMAT_SUBTYPE_PCM,        0x00000001, 0x0000, 0x0010, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71 );
-__DEFINE_GUID(pa_KSDATAFORMAT_SUBTYPE_ADPCM,      0x00000002, 0x0000, 0x0010, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71 );
-__DEFINE_GUID(pa_KSDATAFORMAT_SUBTYPE_IEEE_FLOAT, 0x00000003, 0x0000, 0x0010, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71 );
+__DEFINE_GUID(pa_KSDATAFORMAT_SUBTYPE_PCM,                         0x00000001, 0x0000, 0x0010, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71 );
+__DEFINE_GUID(pa_KSDATAFORMAT_SUBTYPE_ADPCM,                       0x00000002, 0x0000, 0x0010, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71 );
+__DEFINE_GUID(pa_KSDATAFORMAT_SUBTYPE_IEEE_FLOAT,                  0x00000003, 0x0000, 0x0010, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71 );
+__DEFINE_GUID(pa_KSDATAFORMAT_SUBTYPE_IEC61937_PCM,                0x00000000, 0x0000, 0x0010, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71 );
+
+#ifndef _WAVEFORMATEXTENSIBLE_IEC61937_
+    #define _WAVEFORMATEXTENSIBLE_IEC61937_
+typedef struct {
+    WAVEFORMATEXTENSIBLE FormatExt;
+    DWORD                dwEncodedSamplesPerSec;
+    DWORD                dwEncodedChannelCount;
+    DWORD                dwAverageBytesPerSec;
+} WAVEFORMATEXTENSIBLE_IEC61937, *PWAVEFORMATEXTENSIBLE_IEC61937;
+#endif // !_WAVEFORMATEXTENSIBLE_IEC61937_
+
+typedef union _WAVEFORMATEXTENSIBLE_UNION
+{
+    WAVEFORMATEXTENSIBLE          ext;
+    WAVEFORMATEXTENSIBLE_IEC61937 iec61937;
+} WAVEFORMATEXTENSIBLE_UNION;
 
 #ifdef __IAudioClient2_INTERFACE_DEFINED__
 typedef enum _pa_AUDCLNT_STREAMOPTIONS {
@@ -365,6 +383,7 @@ enum { WASAPI_PACKETS_PER_INPUT_BUFFER = 6 };
 
 #define SAFE_CLOSE(h) if ((h) != NULL) { CloseHandle((h)); (h) = NULL; }
 #define SAFE_RELEASE(punk) if ((punk) != NULL) { (punk)->lpVtbl->Release((punk)); (punk) = NULL; }
+#define SAFE_ADDREF(punk) if ((punk) != NULL) { (punk)->lpVtbl->AddRef((punk)); }
 
 // Mixer function
 typedef void (*MixMonoToStereoF) (void *__to, const void *__from, UINT32 count);
@@ -476,12 +495,15 @@ typedef struct PaWasapiDeviceInfo
 
     // Form-factor
     EndpointFormFactor formFactor;
+
+    // Loopback state (TRUE if device acts as loopback)
+    BOOL loopBack;
 }
 PaWasapiDeviceInfo;
 
 // ------------------------------------------------------------------------------------------
 /* PaWasapiHostApiRepresentation - host api datastructure specific to this implementation */
-typedef struct
+typedef struct PaWasapiHostApiRepresentation
 {
     PaUtilHostApiRepresentation inheritedHostApiRep;
     PaUtilStreamInterface       callbackStreamInterface;
@@ -528,7 +550,7 @@ typedef struct PaWasapiSubStream
 #endif
     IAudioClient        *clientProc;
 
-    WAVEFORMATEXTENSIBLE wavex;
+    WAVEFORMATEXTENSIBLE_UNION wavexu;
     UINT32               bufferSize;
     REFERENCE_TIME       deviceLatency;
     REFERENCE_TIME       period;
@@ -567,7 +589,6 @@ PaWasapiHostProcessor;
 // ------------------------------------------------------------------------------------------
 typedef struct PaWasapiStream
 {
-    /* IMPLEMENT ME: rename this */
     PaUtilStreamRepresentation streamRepresentation;
     PaUtilCpuLoadMeasurer      cpuLoadMeasurer;
     PaUtilBufferProcessor      bufferProcessor;
@@ -579,7 +600,6 @@ typedef struct PaWasapiStream
     IStream                   *captureClientStream;
 #endif
     IAudioCaptureClient       *captureClient;
-    IAudioEndpointVolume      *inVol;
 
     // output
     PaWasapiSubStream          out;
@@ -588,23 +608,25 @@ typedef struct PaWasapiStream
     IStream                   *renderClientStream;
 #endif
     IAudioRenderClient        *renderClient;
-    IAudioEndpointVolume      *outVol;
 
-    // event handles for event-driven processing mode
+    // Event handles for event-driven processing mode
     HANDLE event[S_COUNT];
 
-    // buffer mode
+    // Buffer mode
     PaUtilHostBufferSizeMode bufferMode;
 
-    // must be volatile to avoid race condition on user query while
-    // thread is being started
-    volatile BOOL running;
+    // Stream state: active (can be reset inside the processing thread,
+    // to avoid reading incorrected cached state)
+    volatile BOOL isActive;
+
+    // Stream state: stopped (triggered by user only via external API)
+    BOOL isStopped;
 
     PA_THREAD_ID dwThreadId;
     HANDLE hThread;
     HANDLE hCloseRequest;
-    HANDLE hThreadStart;        //!< signalled by thread on start
-    HANDLE hThreadExit;         //!< signalled by thread on exit
+    HANDLE hThreadStart;        // signalled by thread on start
+    HANDLE hThreadExit;         // signalled by thread on exit
     HANDLE hBlockingOpStreamRD;
     HANDLE hBlockingOpStreamWR;
 
@@ -615,7 +637,7 @@ typedef struct PaWasapiStream
     PaWasapiHostProcessor hostProcessOverrideInput;
 
     // Defines blocking/callback interface used
-    BOOL bBlocking;
+    BOOL isBlocking;
 
     // Av Task (MM thread management)
     HANDLE hAvTask;
@@ -639,7 +661,6 @@ static void ReleaseUnmarshaledComPointers(PaWasapiStream *stream);
 
 // Local methods
 static void _StreamOnStop(PaWasapiStream *stream);
-static void _StreamFinish(PaWasapiStream *stream);
 static void _StreamCleanup(PaWasapiStream *stream);
 static HRESULT _PollGetOutputFramesAvailable(PaWasapiStream *stream, UINT32 *available);
 static HRESULT _PollGetInputFramesAvailable(PaWasapiStream *stream, UINT32 *available);
@@ -689,7 +710,7 @@ PaWasapiWinrtDeviceListContext;
 
 // ------------------------------------------------------------------------------------------
 #define LogHostError(HRES) __LogHostError(HRES, __FUNCTION__, __FILE__, __LINE__)
-static HRESULT __LogHostError(HRESULT res, const char *func, const char *file, int line)
+static HRESULT __LogHostError(const HRESULT res, const char *func, const char *file, int line)
 {
     const char *text = NULL;
     switch (res)
@@ -1972,9 +1993,9 @@ static PaError FillDeviceInfo(PaWasapiHostApiRepresentation *paWasapi, void *pEn
 
     // Set default Output/Input devices
     if ((defaultRenderId != NULL) && (wcsncmp(wasapiDeviceInfo->deviceId, defaultRenderId, PA_WASAPI_DEVICE_NAME_LEN - 1) == 0))
-        hostApi->info.defaultOutputDevice = hostApi->info.deviceCount;
+        hostApi->info.defaultOutputDevice = index;
     if ((defaultCaptureId != NULL) && (wcsncmp(wasapiDeviceInfo->deviceId, defaultCaptureId, PA_WASAPI_DEVICE_NAME_LEN - 1) == 0))
-        hostApi->info.defaultInputDevice = hostApi->info.deviceCount;
+        hostApi->info.defaultInputDevice = index;
 
     // Get a temporary IAudioClient for more details
     {
@@ -2068,22 +2089,21 @@ error:
 }
 
 // ------------------------------------------------------------------------------------------
-static PaDeviceInfo *AllocateDeviceListMemory(PaWasapiHostApiRepresentation *paWasapi)
+static PaDeviceInfo *AllocateDeviceListMemory(PaWasapiHostApiRepresentation *paWasapi, UINT32 deviceCount)
 {
     PaUtilHostApiRepresentation *hostApi = (PaUtilHostApiRepresentation *)paWasapi;
     PaDeviceInfo *deviceInfoArray = NULL;
 
     if ((paWasapi->devInfo = (PaWasapiDeviceInfo *)PaUtil_GroupAllocateMemory(paWasapi->allocations,
-        sizeof(PaWasapiDeviceInfo) * paWasapi->deviceCount)) == NULL)
+        sizeof(PaWasapiDeviceInfo) * deviceCount)) == NULL)
     {
         return NULL;
     }
-    memset(paWasapi->devInfo, 0, sizeof(PaWasapiDeviceInfo) * paWasapi->deviceCount);
+    memset(paWasapi->devInfo, 0, sizeof(PaWasapiDeviceInfo) * deviceCount);
 
-    if (paWasapi->deviceCount != 0)
+    if (deviceCount != 0)
     {
         UINT32 i;
-        UINT32 deviceCount = paWasapi->deviceCount;
     #if defined(PA_WASAPI_MAX_CONST_DEVICE_COUNT) && (PA_WASAPI_MAX_CONST_DEVICE_COUNT > 0)
         if (deviceCount < PA_WASAPI_MAX_CONST_DEVICE_COUNT)
             deviceCount = PA_WASAPI_MAX_CONST_DEVICE_COUNT;
@@ -2109,13 +2129,101 @@ static PaDeviceInfo *AllocateDeviceListMemory(PaWasapiHostApiRepresentation *paW
     return deviceInfoArray;
 }
 
+// ------------------------------------------------------------------------------------------
+#ifndef PA_WINRT
+static UINT32 GetDeviceListDeviceCount(IMMDeviceCollection *pEndPoints, EDataFlow filterFlow)
+{
+    HRESULT hr;
+    UINT32 deviceCount;
+    IMMDevice *device;
+    IMMEndpoint *endpoint;
+    EDataFlow flow;
+    UINT32 ret = 0;
+
+    hr = IMMDeviceCollection_GetCount(pEndPoints, &deviceCount);
+    IF_FAILED_JUMP(hr, error);
+
+    for (UINT32 i = 0; i < deviceCount; ++i)
+    {
+        hr = IMMDeviceCollection_Item((IMMDeviceCollection *)pEndPoints, i, &device);
+        IF_FAILED_JUMP(hr, error);
+
+        if (SUCCEEDED(hr = IMMDevice_QueryInterface(device, &pa_IID_IMMEndpoint, (void **)&endpoint)))
+        {
+            if (SUCCEEDED(hr = IMMEndpoint_GetDataFlow(endpoint, &flow)))
+                ret += (flow == filterFlow);
+
+            SAFE_RELEASE(endpoint);
+        }
+
+        SAFE_RELEASE(device);
+    }
+
+    return ret;
+
+error:
+
+    return 0;
+}
+#else
+static UINT32 GetDeviceListDeviceCount(const PaWasapiHostApiRepresentation *paWasapi, 
+    const PaWasapiWinrtDeviceListContext *deviceListContext, EDataFlow filterFlow)
+{
+    UINT32 i, ret = 0;
+
+    for (i = 0; i < paWasapi->deviceCount; ++i)
+        ret += (deviceListContext->devices[i].flow == filterFlow);
+
+    return ret;
+}
+#endif
+
+// ------------------------------------------------------------------------------------------
+static BOOL FillLooopbackDeviceInfo(PaWasapiHostApiRepresentation *paWasapi, PaDeviceInfo *loopbackDeviceInfo,
+    PaWasapiDeviceInfo *loopbackWasapiInfo, const PaDeviceInfo *deviceInfo, const PaWasapiDeviceInfo *wasapiInfo)
+{
+// Loopback device name identificator
+// note: Some projects depend on loopback device detection by device name, do not change!
+#define PA_WASAPI_LOOPBACK_NAME_IDENTIFICATOR "[Loopback]"
+
+    memcpy(loopbackDeviceInfo, deviceInfo, sizeof(*loopbackDeviceInfo));
+    memcpy(loopbackWasapiInfo, wasapiInfo, sizeof(*loopbackWasapiInfo));
+
+    // Append loopback device name identificator to the device name to provide possibility to find
+    // loopback device by its name for some external projects
+    loopbackDeviceInfo->name = (char *)PaUtil_GroupAllocateMemory(paWasapi->allocations, PA_WASAPI_DEVICE_NAME_LEN + 1);
+    if (loopbackDeviceInfo->name == NULL)
+        return FALSE;
+    _snprintf((char *)loopbackDeviceInfo->name, PA_WASAPI_DEVICE_NAME_LEN - 1,
+        "%s " PA_WASAPI_LOOPBACK_NAME_IDENTIFICATOR, deviceInfo->name);
+
+    // Acquire ref to the device as it is already referenced as render (output) device
+    // to avoid duplicate release in ReleaseWasapiDeviceInfoList()
+#ifndef PA_WINRT
+    SAFE_ADDREF(loopbackWasapiInfo->device);
+#endif
+
+    // Mark as loopback device
+    loopbackWasapiInfo->loopBack = TRUE;
+
+    // Input is the reverse of Output
+    loopbackDeviceInfo->maxInputChannels         = deviceInfo->maxOutputChannels;
+    loopbackDeviceInfo->defaultHighInputLatency     = deviceInfo->defaultHighOutputLatency;
+    loopbackDeviceInfo->defaultLowInputLatency     = deviceInfo->defaultLowOutputLatency;
+    loopbackDeviceInfo->maxOutputChannels         = 0;
+    loopbackDeviceInfo->defaultHighOutputLatency = 0;
+    loopbackDeviceInfo->defaultLowOutputLatency     = 0;
+
+    return TRUE;
+}
+
 // ------------------------------------------------------------------------------------------
 static PaError CreateDeviceList(PaWasapiHostApiRepresentation *paWasapi, PaHostApiIndex hostApiIndex)
 {
     PaUtilHostApiRepresentation *hostApi = (PaUtilHostApiRepresentation *)paWasapi;
     PaError result = paNoError;
     PaDeviceInfo *deviceInfoArray = NULL;
-    UINT32 i;
+    UINT32 i, j, loopbackDevices;
     WCHAR *defaultRenderId = NULL;
     WCHAR *defaultCaptureId = NULL;
 #ifndef PA_WINRT
@@ -2181,6 +2289,9 @@ static PaError CreateDeviceList(PaWasapiHostApiRepresentation *paWasapi, PaHostA
     // Get device count
     hr = IMMDeviceCollection_GetCount(pEndPoints, &paWasapi->deviceCount);
     IF_FAILED_INTERNAL_ERROR_JUMP(hr, result, error);
+
+    // Get loopback device count (e.g. all renders)
+    loopbackDevices = GetDeviceListDeviceCount(pEndPoints, eRender);
 #else
     WinRT_GetDefaultDeviceId(defaultRender.id, STATIC_ARRAY_SIZE(defaultRender.id) - 1, eRender);
     defaultRenderId = defaultRender.id;
@@ -2229,19 +2340,24 @@ static PaError CreateDeviceList(PaWasapiHostApiRepresentation *paWasapi, PaHostA
             paWasapi->deviceCount++;
         }
     }
+
+    // Get loopback device count (e.g. all renders)
+    loopbackDevices = GetDeviceListDeviceCount(paWasapi, &deviceListContext, eRender);
 #endif
 
     // Allocate memory for the device list
-    if ((paWasapi->deviceCount != 0) && ((deviceInfoArray = AllocateDeviceListMemory(paWasapi)) == NULL))
+    if ((paWasapi->deviceCount != 0) &&
+        ((deviceInfoArray = AllocateDeviceListMemory(paWasapi, paWasapi->deviceCount + loopbackDevices)) == NULL))
     {
         result = paInsufficientMemory;
         goto error;
     }
 
     // Fill WASAPI device info
-    for (i = 0; i < paWasapi->deviceCount; ++i)
+    for (i = 0, j = 0; i < paWasapi->deviceCount; ++i)
     {
         PaDeviceInfo *deviceInfo = &deviceInfoArray[i];
+        PaWasapiDeviceInfo *wasapiInfo = &paWasapi->devInfo[i];
 
         PA_DEBUG(("WASAPI: device idx: %02d\n", i));
         PA_DEBUG(("WASAPI: ---------------\n"));
@@ -2249,7 +2365,7 @@ static PaError CreateDeviceList(PaWasapiHostApiRepresentation *paWasapi, PaHostA
         FillBaseDeviceInfo(deviceInfo, hostApiIndex);
 
         if ((result = FillDeviceInfo(paWasapi, pEndPoints, i, defaultRenderId, defaultCaptureId,
-            deviceInfo, &paWasapi->devInfo[i]
+            deviceInfo, wasapiInfo
         #ifdef PA_WINRT
             , &deviceListContext
         #endif
@@ -2262,8 +2378,30 @@ static PaError CreateDeviceList(PaWasapiHostApiRepresentation *paWasapi, PaHostA
 
         hostApi->deviceInfos[i] = deviceInfo;
         ++hostApi->info.deviceCount;
+
+        // Add loopback device for the render device
+        if ((paWasapi->devInfo[i].flow == eRender) && (j < loopbackDevices))
+        {
+            // Add loopback device to the end of the device list
+            UINT32 loopbackIndex = paWasapi->deviceCount + j++;
+            assert(loopbackIndex < (paWasapi->deviceCount + loopbackDevices));
+            PaDeviceInfo *loopbackDeviceInfo = &deviceInfoArray[loopbackIndex];
+            PaWasapiDeviceInfo *loopbackWasapiInfo = &paWasapi->devInfo[loopbackIndex];
+
+            if (!FillLooopbackDeviceInfo(paWasapi, loopbackDeviceInfo, loopbackWasapiInfo, deviceInfo, wasapiInfo))
+            {
+                result = paInsufficientMemory;
+                goto error;
+            }
+
+            hostApi->deviceInfos[loopbackIndex] = loopbackDeviceInfo;
+            ++hostApi->info.deviceCount;
+        }
     }
 
+    // Resize device list to accomodate inserted loopback devices
+    paWasapi->deviceCount += loopbackDevices;
+
     // Fill the remaining slots with inactive device info
 #if defined(PA_WASAPI_MAX_CONST_DEVICE_COUNT) && (PA_WASAPI_MAX_CONST_DEVICE_COUNT > 0)
     if ((hostApi->info.deviceCount != 0) && (hostApi->info.deviceCount < PA_WASAPI_MAX_CONST_DEVICE_COUNT))
@@ -2286,7 +2424,7 @@ static PaError CreateDeviceList(PaWasapiHostApiRepresentation *paWasapi, PaHostA
     // Clear any non-fatal errors
     result = paNoError;
 
-    PRINT(("WASAPI: device list ok - found %d devices\n", paWasapi->deviceCount));
+    PRINT(("WASAPI: device list ok - found %d devices, %d loopback devices\n", paWasapi->deviceCount, loopbackDevices));
 
 done:
 
@@ -2521,7 +2659,7 @@ int PaWasapi_GetDeviceCurrentFormat( PaStream *pStream, void *pFormat, unsigned
     if (stream == NULL)
         return paBadStreamPtr;
 
-    format = (bOutput == TRUE ? &stream->out.wavex : &stream->in.wavex);
+    format = (bOutput == TRUE ? &stream->out.wavexu.ext : &stream->in.wavexu.ext);
 
     size = min(formatSize, (UINT32)sizeof(*format));
     memcpy(pFormat, format, size);
@@ -2630,6 +2768,29 @@ PaError PaWasapi_GetIMMDevice( PaDeviceIndex device, void **pIMMDevice )
 #endif
 }
 
+// ------------------------------------------------------------------------------------------
+int PaWasapi_IsLoopback( PaDeviceIndex device )
+{
+    PaError ret;
+    PaDeviceIndex index;
+
+    // Get API
+    PaWasapiHostApiRepresentation *paWasapi = _GetHostApi(&ret);
+    if (paWasapi == NULL)
+        return paNotInitialized;
+
+    // Get device index
+    ret = PaUtil_DeviceIndexToHostApiDeviceIndex(&index, device, &paWasapi->inheritedHostApiRep);
+    if (ret != paNoError)
+        return ret;
+
+    // Validate index
+    if ((UINT32)index >= paWasapi->deviceCount)
+        return paInvalidDevice;
+
+    return paWasapi->devInfo[ index ].loopBack;
+}
+
 // ------------------------------------------------------------------------------------------
 PaError PaWasapi_GetFramesPerHostBuffer( PaStream *pStream, unsigned int *pInput, unsigned int *pOutput )
 {
@@ -2702,9 +2863,7 @@ static void LogWAVEFORMATEXTENSIBLE(const WAVEFORMATEXTENSIBLE *in)
 // ------------------------------------------------------------------------------------------
 PaSampleFormat WaveToPaFormat(const WAVEFORMATEXTENSIBLE *fmtext)
 {
-    const WAVEFORMATEX *fmt = (WAVEFORMATEX *)fmtext;
-
-    switch (fmt->wFormatTag)
+    switch (fmtext->Format.wFormatTag)
     {
     case WAVE_FORMAT_EXTENSIBLE: {
         if (IsEqualGUID(&fmtext->SubFormat, &pa_KSDATAFORMAT_SUBTYPE_IEEE_FLOAT))
@@ -2715,7 +2874,7 @@ PaSampleFormat WaveToPaFormat(const WAVEFORMATEXTENSIBLE *fmtext)
         else
         if (IsEqualGUID(&fmtext->SubFormat, &pa_KSDATAFORMAT_SUBTYPE_PCM))
         {
-            switch (fmt->wBitsPerSample)
+            switch (fmtext->Format.wBitsPerSample)
             {
             case 32: return paInt32;
             case 24: return paInt24;
@@ -2723,13 +2882,25 @@ PaSampleFormat WaveToPaFormat(const WAVEFORMATEXTENSIBLE *fmtext)
             case  8: return paUInt8;
             }
         }
+        else
+        // KSDATAFORMAT_SUBTYPE_IEC61937_PCM is a naked format GUID which has different Data1 and Data2
+        // depending on the passthrough format, for the possible values refer Microsoft documentation
+        // "Representing Formats for IEC 61937 Transmissions". Here we check if Data3 and Data4 match
+        // assuming that if they do match then we have KSDATAFORMAT_SUBTYPE_IEC61937_XXX format.
+        // Currently PA WASAPI is using KSDATAFORMAT_SUBTYPE_IEEE_FLOAT and KSDATAFORMAT_SUBTYPE_PCM
+        // therefore this check is reliable in this way.
+        if (!memcmp(&fmtext->SubFormat.Data3, &pa_KSDATAFORMAT_SUBTYPE_IEC61937_PCM.Data3,
+            (sizeof(GUID) - offsetof(GUID, Data3))))
+        {
+            return paInt16;
+        }
         break; }
 
     case WAVE_FORMAT_IEEE_FLOAT:
         return paFloat32;
 
     case WAVE_FORMAT_PCM: {
-        switch (fmt->wBitsPerSample)
+        switch (fmtext->Format.wBitsPerSample)
         {
         case 32: return paInt32;
         case 24: return paInt24;
@@ -2743,7 +2914,7 @@ PaSampleFormat WaveToPaFormat(const WAVEFORMATEXTENSIBLE *fmtext)
 }
 
 // ------------------------------------------------------------------------------------------
-static PaError MakeWaveFormatFromParams(WAVEFORMATEXTENSIBLE *wavex, const PaStreamParameters *params,
+static PaError MakeWaveFormatFromParams(WAVEFORMATEXTENSIBLE_UNION *wavexu, const PaStreamParameters *params,
     double sampleRate, BOOL packedOnly)
 {
     WORD bitsPerSample;
@@ -2751,6 +2922,8 @@ static PaError MakeWaveFormatFromParams(WAVEFORMATEXTENSIBLE *wavex, const PaStr
     DWORD channelMask = 0;
     BOOL useExtensible = (params->channelCount > 2); // format is always forced for >2 channels format
     PaWasapiStreamInfo *streamInfo = (PaWasapiStreamInfo *)params->hostApiSpecificStreamInfo;
+    WAVEFORMATEXTENSIBLE *wavex = (WAVEFORMATEXTENSIBLE *)wavexu;
+    BOOL passthroughMode = ((streamInfo != NULL) && (streamInfo->flags & paWinWasapiPassthrough));
 
     // Convert PaSampleFormat to valid data bits
     if ((bitsPerSample = PaSampleFormatToBitsPerSample(params->sampleFormat)) == 0)
@@ -2763,9 +2936,9 @@ static PaError MakeWaveFormatFromParams(WAVEFORMATEXTENSIBLE *wavex, const PaStr
         useExtensible = TRUE;
     }
 
-    memset(wavex, 0, sizeof(*wavex));
+    memset(wavexu, 0, sizeof(*wavexu));
 
-    old                    = (WAVEFORMATEX *)wavex;
+    old                 = (WAVEFORMATEX *)wavex;
     old->nChannels      = (WORD)params->channelCount;
     old->nSamplesPerSec = (DWORD)sampleRate;
     old->wBitsPerSample = bitsPerSample;
@@ -2783,18 +2956,35 @@ static PaError MakeWaveFormatFromParams(WAVEFORMATEXTENSIBLE *wavex, const PaStr
     // WAVEFORMATEX
     if (!useExtensible)
     {
-        old->wFormatTag    = WAVE_FORMAT_PCM;
+        old->wFormatTag = WAVE_FORMAT_PCM;
     }
     // WAVEFORMATEXTENSIBLE
     else
     {
         old->wFormatTag = WAVE_FORMAT_EXTENSIBLE;
-        old->cbSize        = sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX);
+        old->cbSize = (passthroughMode ? (sizeof(WAVEFORMATEXTENSIBLE_IEC61937) - sizeof(WAVEFORMATEX))
+            : (sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX)));
+
+        if (passthroughMode)
+        {
+            WAVEFORMATEXTENSIBLE_IEC61937 *wavex_61937 = (WAVEFORMATEXTENSIBLE_IEC61937 *)wavexu;
 
-        if ((params->sampleFormat & ~paNonInterleaved) == paFloat32)
+            wavex->SubFormat = pa_KSDATAFORMAT_SUBTYPE_IEC61937_PCM;
+            wavex->SubFormat.Data1 = (streamInfo->passthrough.formatId >> 16);
+            wavex->SubFormat.Data2 = (USHORT)(streamInfo->passthrough.formatId & 0x0000FFFF);
+
+            wavex_61937->dwEncodedSamplesPerSec = streamInfo->passthrough.encodedSamplesPerSec;
+            wavex_61937->dwEncodedChannelCount = streamInfo->passthrough.encodedChannelCount;
+            wavex_61937->dwAverageBytesPerSec = streamInfo->passthrough.averageBytesPerSec;
+        }
+        else if ((params->sampleFormat & ~paNonInterleaved) == paFloat32)
+        {
             wavex->SubFormat = pa_KSDATAFORMAT_SUBTYPE_IEEE_FLOAT;
+        }
         else
+        {
             wavex->SubFormat = pa_KSDATAFORMAT_SUBTYPE_PCM;
+        }
 
         wavex->Samples.wValidBitsPerSample = bitsPerSample;
 
@@ -2841,11 +3031,11 @@ static PaError MakeWaveFormatFromParams(WAVEFORMATEXTENSIBLE *wavex, const PaStr
 
 // ------------------------------------------------------------------------------------------
 static HRESULT GetAlternativeSampleFormatExclusive(IAudioClient *client, double sampleRate,
-    const PaStreamParameters *params, WAVEFORMATEXTENSIBLE *outWavex, BOOL packedSampleFormatOnly)
+    const PaStreamParameters *params, WAVEFORMATEXTENSIBLE_UNION *outWavexU, BOOL packedSampleFormatOnly)
 {
     HRESULT hr = !S_OK;
     AUDCLNT_SHAREMODE shareMode = AUDCLNT_SHAREMODE_EXCLUSIVE;
-    WAVEFORMATEXTENSIBLE testFormat;
+    WAVEFORMATEXTENSIBLE_UNION testFormat;
     PaStreamParameters testParams;
     int i;
     static const PaSampleFormat bestToWorst[] = { paInt32, paInt24, paFloat32, paInt16 };
@@ -2858,9 +3048,9 @@ static HRESULT GetAlternativeSampleFormatExclusive(IAudioClient *client, double
 
         if (MakeWaveFormatFromParams(&testFormat, &testParams, sampleRate, packedSampleFormatOnly) == paNoError)
         {
-            if ((hr = IAudioClient_IsFormatSupported(client, shareMode, &testFormat.Format, NULL)) == S_OK)
+            if ((hr = IAudioClient_IsFormatSupported(client, shareMode, &testFormat.ext.Format, NULL)) == S_OK)
             {
-                (*outWavex) = testFormat;
+                (*outWavexU) = testFormat;
                 return hr;
             }
         }
@@ -2872,9 +3062,9 @@ static HRESULT GetAlternativeSampleFormatExclusive(IAudioClient *client, double
 
             if (MakeWaveFormatFromParams(&testFormat, &testParams, sampleRate, packedSampleFormatOnly) == paNoError)
             {
-                if ((hr = IAudioClient_IsFormatSupported(client, shareMode, &testFormat.Format, NULL)) == S_OK)
+                if ((hr = IAudioClient_IsFormatSupported(client, shareMode, &testFormat.ext.Format, NULL)) == S_OK)
                 {
-                    (*outWavex) = testFormat;
+                    (*outWavexU) = testFormat;
                     return hr;
                 }
             }
@@ -2889,9 +3079,9 @@ static HRESULT GetAlternativeSampleFormatExclusive(IAudioClient *client, double
 
         if (MakeWaveFormatFromParams(&testFormat, &testParams, sampleRate, packedSampleFormatOnly) == paNoError)
         {
-            if ((hr = IAudioClient_IsFormatSupported(client, shareMode, &testFormat.Format, NULL)) == S_OK)
+            if ((hr = IAudioClient_IsFormatSupported(client, shareMode, &testFormat.ext.Format, NULL)) == S_OK)
             {
-                (*outWavex) = testFormat;
+                (*outWavexU) = testFormat;
                 return hr;
             }
         }
@@ -2902,13 +3092,14 @@ static HRESULT GetAlternativeSampleFormatExclusive(IAudioClient *client, double
 
 // ------------------------------------------------------------------------------------------
 static PaError GetClosestFormat(IAudioClient *client, double sampleRate, const PaStreamParameters *_params,
-    AUDCLNT_SHAREMODE shareMode, WAVEFORMATEXTENSIBLE *outWavex, BOOL output)
+    AUDCLNT_SHAREMODE shareMode, WAVEFORMATEXTENSIBLE_UNION *outWavexU, BOOL output)
 {
     PaWasapiStreamInfo *streamInfo   = (PaWasapiStreamInfo *)_params->hostApiSpecificStreamInfo;
     WAVEFORMATEX *sharedClosestMatch = NULL;
     HRESULT hr                       = !S_OK;
     PaStreamParameters params        = (*_params);
     const BOOL explicitFormat        = (streamInfo != NULL) && ((streamInfo->flags & paWinWasapiExplicitSampleFormat) == paWinWasapiExplicitSampleFormat);
+    WAVEFORMATEXTENSIBLE *outWavex   = (WAVEFORMATEXTENSIBLE *)outWavexU;
     (void)output;
 
     /* It was not noticed that 24-bit Input producing no output while device accepts this format.
@@ -2920,7 +3111,7 @@ static PaError GetClosestFormat(IAudioClient *client, double sampleRate, const P
         params.sampleFormat = paFloat32;*/ // <<< The silence was due to missing Int32_To_Int24_Dither implementation
 
     // Try standard approach, e.g. if data is > 16 bits it will be packed into 32-bit containers
-    MakeWaveFormatFromParams(outWavex, &params, sampleRate, FALSE);
+    MakeWaveFormatFromParams(outWavexU, &params, sampleRate, FALSE);
 
     // If built-in PCM converter requested then shared mode format will always succeed
     if ((GetWindowsVersion() >= WINDOWS_7_SERVER2008R2) &&
@@ -2934,7 +3125,7 @@ static PaError GetClosestFormat(IAudioClient *client, double sampleRate, const P
     if ((hr != S_OK) && (shareMode == AUDCLNT_SHAREMODE_EXCLUSIVE))
     {
         // Enforce packed only format, e.g. data bits will not be packed into 32-bit containers in any case
-        MakeWaveFormatFromParams(outWavex, &params, sampleRate, TRUE);
+        MakeWaveFormatFromParams(outWavexU, &params, sampleRate, TRUE);
         hr = IAudioClient_IsFormatSupported(client, shareMode, &outWavex->Format, NULL);
     }
 
@@ -2982,11 +3173,11 @@ static PaError GetClosestFormat(IAudioClient *client, double sampleRate, const P
     if ((shareMode == AUDCLNT_SHAREMODE_EXCLUSIVE) && !explicitFormat)
     {
         // Try standard approach, e.g. if data is > 16 bits it will be packed into 32-bit containers
-        if ((hr = GetAlternativeSampleFormatExclusive(client, sampleRate, &params, outWavex, FALSE)) == S_OK)
+        if ((hr = GetAlternativeSampleFormatExclusive(client, sampleRate, &params, outWavexU, FALSE)) == S_OK)
             return paFormatIsSupported;
 
         // Enforce packed only format, e.g. data bits will not be packed into 32-bit containers in any case
-        if ((hr = GetAlternativeSampleFormatExclusive(client, sampleRate, &params, outWavex, TRUE)) == S_OK)
+        if ((hr = GetAlternativeSampleFormatExclusive(client, sampleRate, &params, outWavexU, TRUE)) == S_OK)
             return paFormatIsSupported;
 
         // Log failure
@@ -3096,7 +3287,7 @@ static PaError IsFormatSupported( struct PaUtilHostApiRepresentation *hostApi,
 
     if (inputParameters != NULL)
     {
-        WAVEFORMATEXTENSIBLE wavex;
+        WAVEFORMATEXTENSIBLE_UNION wavex;
         HRESULT hr;
         PaError answer;
         AUDCLNT_SHAREMODE shareMode = AUDCLNT_SHAREMODE_SHARED;
@@ -3122,7 +3313,7 @@ static PaError IsFormatSupported( struct PaUtilHostApiRepresentation *hostApi,
     if (outputParameters != NULL)
     {
         HRESULT hr;
-        WAVEFORMATEXTENSIBLE wavex;
+        WAVEFORMATEXTENSIBLE_UNION wavex;
         PaError answer;
         AUDCLNT_SHAREMODE shareMode = AUDCLNT_SHAREMODE_SHARED;
         outputStreamInfo = (PaWasapiStreamInfo *)outputParameters->hostApiSpecificStreamInfo;
@@ -3148,11 +3339,14 @@ static PaError IsFormatSupported( struct PaUtilHostApiRepresentation *hostApi,
 }
 
 // ------------------------------------------------------------------------------------------
-static PaUint32 _GetFramesPerHostBuffer(PaUint32 userFramesPerBuffer, PaTime suggestedLatency, double sampleRate, PaUint32 TimerJitterMs)
+static UINT32 _CalculateFramesPerHostBuffer(UINT32 userFramesPerBuffer, PaTime suggestedLatency, double sampleRate)
 {
-    PaUint32 frames = userFramesPerBuffer + max( userFramesPerBuffer, (PaUint32)(suggestedLatency * sampleRate) );
-    frames += (PaUint32)((sampleRate * 0.001) * TimerJitterMs);
-    return frames;
+    // Note: WASAPI tends to enforce double buffering by itself therefore enforcing double buffering
+    // on PA side looks as an excessive measure.
+    const BOOL doubleBuffering = TRUE;
+
+    return userFramesPerBuffer + max((doubleBuffering ? userFramesPerBuffer : 0),
+        (UINT32)(suggestedLatency * sampleRate));
 }
 
 // ------------------------------------------------------------------------------------------
@@ -3167,7 +3361,7 @@ static void _RecalculateBuffersCount(PaWasapiSubStream *sub, UINT32 userFramesPe
     // Determine number of buffers used:
     // - Full-duplex mode will lead to period difference, thus only 1
     // - Input mode, only 1, as WASAPI allows extraction of only 1 packet
-    // - For Shared mode we use double buffering
+    // - For Shared mode we use double buffering (see _CalculateFramesPerHostBuffer)
     if ((sub->shareMode == AUDCLNT_SHAREMODE_EXCLUSIVE) || fullDuplex)
     {
         BOOL eventMode = ((sub->streamFlags & AUDCLNT_STREAMFLAGS_EVENTCALLBACK) == AUDCLNT_STREAMFLAGS_EVENTCALLBACK);
@@ -3175,7 +3369,7 @@ static void _RecalculateBuffersCount(PaWasapiSubStream *sub, UINT32 userFramesPe
         // Exclusive mode does not allow >1 buffers be used for Event interface, e.g. GetBuffer
         // call must acquire max buffer size and it all must be processed.
         if (eventMode)
-            sub->userBufferAndHostMatch = 1;
+            sub->userBufferAndHostMatch = TRUE;
 
         // Full-duplex or Event mode: prefer paUtilBoundedHostBufferSize because exclusive mode will starve
         // and produce glitchy audio
@@ -3196,11 +3390,11 @@ static void _CalculateAlignedPeriod(PaWasapiSubStream *pSub, UINT32 *nFramesPerL
     if (pSub->shareMode == AUDCLNT_SHAREMODE_EXCLUSIVE)
     {
         (*nFramesPerLatency) = AlignFramesPerBuffer((*nFramesPerLatency),
-            pSub->wavex.Format.nBlockAlign, pAlignFunc);
+            pSub->wavexu.ext.Format.nBlockAlign, pAlignFunc);
     }
 
     // Calculate period
-    pSub->period = MakeHnsPeriod((*nFramesPerLatency), pSub->wavex.Format.nSamplesPerSec);
+    pSub->period = MakeHnsPeriod((*nFramesPerLatency), pSub->wavexu.ext.Format.nSamplesPerSec);
 }
 
 // ------------------------------------------------------------------------------------------
@@ -3231,9 +3425,9 @@ static void _CalculatePeriodicity(PaWasapiSubStream *pSub, BOOL output, REFERENC
             // Align frames backwards, so device will likely make buffer read ready when we are ready
             // to read it (our scheduling will wait for amount of millisoconds of frames_per_buffer)
             alignedFrames = AlignFramesPerBuffer(pSub->params.frames_per_buffer,
-                pSub->wavex.Format.nBlockAlign, ALIGN_BWD);
+                pSub->wavexu.ext.Format.nBlockAlign, ALIGN_BWD);
 
-            userPeriodicity = MakeHnsPeriod(alignedFrames, pSub->wavex.Format.nSamplesPerSec);
+            userPeriodicity = MakeHnsPeriod(alignedFrames, pSub->wavexu.ext.Format.nSamplesPerSec);
 
             // Must not be larger than buffer size
             if (userPeriodicity > pSub->period)
@@ -3288,7 +3482,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
     }
 
     // Get closest format
-    if ((error = GetClosestFormat(audioClient, sampleRate, params, pSub->shareMode, &pSub->wavex, output)) != paFormatIsSupported)
+    if ((error = GetClosestFormat(audioClient, sampleRate, params, pSub->shareMode, &pSub->wavexu, output)) != paFormatIsSupported)
     {
         (*pa_error) = error;
         LogHostError(hr = AUDCLNT_E_UNSUPPORTED_FORMAT);
@@ -3296,10 +3490,10 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
     }
 
     // Check for Mono <<>> Stereo workaround
-    if ((params->channelCount == 1) && (pSub->wavex.Format.nChannels == 2))
+    if ((params->channelCount == 1) && (pSub->wavexu.ext.Format.nChannels == 2))
     {
         // select mixer
-        pSub->monoMixer = GetMonoToStereoMixer(&pSub->wavex, (pInfo->flow == eRender ? MIX_DIR__1TO2 : MIX_DIR__2TO1_L));
+        pSub->monoMixer = GetMonoToStereoMixer(&pSub->wavexu.ext, (output ? MIX_DIR__1TO2 : MIX_DIR__2TO1_L));
         if (pSub->monoMixer == NULL)
         {
             (*pa_error) = paInvalidChannelCount;
@@ -3312,9 +3506,8 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
     if ((pSub->shareMode != AUDCLNT_SHAREMODE_EXCLUSIVE) &&
         (!pSub->streamFlags || ((pSub->streamFlags & AUDCLNT_STREAMFLAGS_EVENTCALLBACK) == 0)))
     {
-        framesPerLatency = _GetFramesPerHostBuffer(userFramesPerBuffer,
-            params->suggestedLatency, pSub->wavex.Format.nSamplesPerSec, 0/*,
-            (pSub->streamFlags & AUDCLNT_STREAMFLAGS_EVENTCALLBACK ? 0 : 1)*/);
+        framesPerLatency = _CalculateFramesPerHostBuffer(userFramesPerBuffer, params->suggestedLatency,
+            pSub->wavexu.ext.Format.nSamplesPerSec);
     }
     else
     {
@@ -3323,17 +3516,16 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
     #endif
 
         // Work 1:1 with user buffer (only polling allows to use >1)
-        framesPerLatency += MakeFramesFromHns(SecondsTonano100(params->suggestedLatency), pSub->wavex.Format.nSamplesPerSec);
+        framesPerLatency += MakeFramesFromHns(SecondsTonano100(params->suggestedLatency), pSub->wavexu.ext.Format.nSamplesPerSec);
 
         // Force Polling if overall latency is >= 21.33ms as it allows to use 100% CPU in a callback,
         // or user specified latency parameter.
     #ifdef PA_WASAPI_FORCE_POLL_IF_LARGE_BUFFER
-        overall = MakeHnsPeriod(framesPerLatency, pSub->wavex.Format.nSamplesPerSec);
+        overall = MakeHnsPeriod(framesPerLatency, pSub->wavexu.ext.Format.nSamplesPerSec);
         if (overall >= (106667 * 2)/*21.33ms*/)
         {
-            framesPerLatency = _GetFramesPerHostBuffer(userFramesPerBuffer,
-                params->suggestedLatency, pSub->wavex.Format.nSamplesPerSec, 0/*,
-                (streamFlags & AUDCLNT_STREAMFLAGS_EVENTCALLBACK ? 0 : 1)*/);
+            framesPerLatency = _CalculateFramesPerHostBuffer(userFramesPerBuffer, params->suggestedLatency,
+                pSub->wavexu.Format.nSamplesPerSec);
 
             // Use Polling interface
             pSub->streamFlags &= ~AUDCLNT_STREAMFLAGS_EVENTCALLBACK;
@@ -3348,7 +3540,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
 
     // Avoid 0 frames
     if (framesPerLatency == 0)
-        framesPerLatency = MakeFramesFromHns(pInfo->DefaultDevicePeriod, pSub->wavex.Format.nSamplesPerSec);
+        framesPerLatency = MakeFramesFromHns(pInfo->DefaultDevicePeriod, pSub->wavexu.ext.Format.nSamplesPerSec);
 
     // Exclusive Input stream renders data in 6 packets, we must set then the size of
     // single packet, total buffer size, e.g. required latency will be PacketSize * 6
@@ -3372,7 +3564,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
             pSub->period = pInfo->DefaultDevicePeriod;
 
             // Recalculate aligned period
-            framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavex.Format.nSamplesPerSec);
+            framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavexu.ext.Format.nSamplesPerSec);
             _CalculateAlignedPeriod(pSub, &framesPerLatency, ALIGN_BWD);
         }
     }
@@ -3383,7 +3575,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
             pSub->period = pInfo->MinimumDevicePeriod;
 
             // Recalculate aligned period
-            framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavex.Format.nSamplesPerSec);
+            framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavexu.ext.Format.nSamplesPerSec);
             _CalculateAlignedPeriod(pSub, &framesPerLatency, ALIGN_FWD);
         }
     }
@@ -3411,7 +3603,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
                     pSub->period = MAX_BUFFER_EVENT_DURATION;
 
                     // Recalculate aligned period
-                    framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavex.Format.nSamplesPerSec);
+                    framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavexu.ext.Format.nSamplesPerSec);
                     _CalculateAlignedPeriod(pSub, &framesPerLatency, ALIGN_BWD);
                 }
             }
@@ -3423,7 +3615,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
                     pSub->period = MAX_BUFFER_POLL_DURATION;
 
                     // Recalculate aligned period
-                    framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavex.Format.nSamplesPerSec);
+                    framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavexu.ext.Format.nSamplesPerSec);
                     _CalculateAlignedPeriod(pSub, &framesPerLatency, ALIGN_BWD);
                 }
             }
@@ -3439,7 +3631,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
         pSub->streamFlags,
         pSub->period,
         eventPeriodicity,
-        &pSub->wavex.Format,
+        &pSub->wavexu.ext.Format,
         NULL);
 
     // [Output only] Check if buffer size is the one we requested in Exclusive mode, for UAC1 USB DACs WASAPI
@@ -3466,7 +3658,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
             PRINT(("WASAPI: CreateAudioClient: detected %d times larger buffer than requested, correct to match user latency\n", ratio));
 
             // Get new aligned frames lowered by calculated ratio
-            framesPerLatency = MakeFramesFromHns(pSub->period / ratio, pSub->wavex.Format.nSamplesPerSec);
+            framesPerLatency = MakeFramesFromHns(pSub->period / ratio, pSub->wavexu.ext.Format.nSamplesPerSec);
             _CalculateAlignedPeriod(pSub, &framesPerLatency, ALIGN_BWD);
 
             // Make sure we are not below the minimum period
@@ -3493,7 +3685,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
                 pSub->streamFlags,
                 pSub->period,
                 eventPeriodicity,
-                &pSub->wavex.Format,
+                &pSub->wavexu.ext.Format,
                 NULL);
         }
     }
@@ -3510,7 +3702,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
         pSub->period -= (100 * 10000);
 
         // Recalculate aligned period
-        framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavex.Format.nSamplesPerSec);
+        framesPerLatency = MakeFramesFromHns(pSub->period, pSub->wavexu.ext.Format.nSamplesPerSec);
         _CalculateAlignedPeriod(pSub, &framesPerLatency, ALIGN_BWD);
 
         // Release the previous allocations
@@ -3533,7 +3725,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
             pSub->streamFlags,
             pSub->period,
             eventPeriodicity,
-            &pSub->wavex.Format,
+            &pSub->wavexu.ext.Format,
             NULL);
     }
 
@@ -3566,7 +3758,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
             pSub->streamFlags,
             pSub->period,
             eventPeriodicity,
-            &pSub->wavex.Format,
+            &pSub->wavexu.ext.Format,
             NULL);
     }
 
@@ -3583,7 +3775,7 @@ static HRESULT CreateAudioClient(PaWasapiStream *pStream, PaWasapiSubStream *pSu
     IAudioClient_AddRef(pSub->clientParent);
 
     // Recalculate buffers count
-    _RecalculateBuffersCount(pSub, userFramesPerBuffer, MakeFramesFromHns(pSub->period, pSub->wavex.Format.nSamplesPerSec),
+    _RecalculateBuffersCount(pSub, userFramesPerBuffer, MakeFramesFromHns(pSub->period, pSub->wavexu.ext.Format.nSamplesPerSec),
         fullDuplex, output);
 
     // No error, client is successfully created
@@ -3597,90 +3789,24 @@ done:
 }
 
 // ------------------------------------------------------------------------------------------
-static PaError ActivateAudioClientOutput(PaWasapiStream *stream)
-{
-    HRESULT hr;
-    PaError result;
-    UINT32 maxBufferSize;
-    PaTime bufferLatency;
-    const UINT32 framesPerBuffer = stream->out.params.frames_per_buffer;
-
-    // Create Audio client
-    if (FAILED(hr = CreateAudioClient(stream, &stream->out, TRUE, &result)))
-    {
-        LogPaError(result);
-        goto error;
-    }
-    LogWAVEFORMATEXTENSIBLE(&stream->out.wavex);
-
-    // Activate volume
-    stream->outVol = NULL;
-    /*hr = info->device->Activate(
-        __uuidof(IAudioEndpointVolume), CLSCTX_INPROC_SERVER, NULL,
-        (void**)&stream->outVol);
-    if (hr != S_OK)
-        return paInvalidDevice;*/
-
-    // Get max possible buffer size to check if it is not less than that we request
-    if (FAILED(hr = IAudioClient_GetBufferSize(stream->out.clientParent, &maxBufferSize)))
-    {
-        LogHostError(hr);
-        LogPaError(result = paInvalidDevice);
-        goto error;
-    }
-
-    // Correct buffer to max size if it maxed out result of GetBufferSize
-    stream->out.bufferSize = maxBufferSize;
-
-    // Number of frames that are required at each period
-    stream->out.framesPerHostCallback = maxBufferSize;
-
-    // Calculate frames per single buffer, if buffers > 1 then always framesPerBuffer
-    stream->out.framesPerBuffer =
-        (stream->out.userBufferAndHostMatch ? stream->out.framesPerHostCallback : framesPerBuffer);
-
-    // Calculate buffer latency
-    bufferLatency = (PaTime)maxBufferSize / stream->out.wavex.Format.nSamplesPerSec;
-
-    // Append buffer latency to interface latency in shared mode (see GetStreamLatency notes)
-    stream->out.latencySeconds = bufferLatency;
-
-    PRINT(("WASAPI::OpenStream(output): framesPerUser[ %d ] framesPerHost[ %d ] latency[ %.02fms ] exclusive[ %s ] wow64_fix[ %s ] mode[ %s ]\n", (UINT32)framesPerBuffer, (UINT32)stream->out.framesPerHostCallback, (float)(stream->out.latencySeconds*1000.0f), (stream->out.shareMode == AUDCLNT_SHAREMODE_EXCLUSIVE ? "YES" : "NO"), (stream->out.params.wow64_workaround ? "YES" : "NO"), (stream->out.streamFlags & AUDCLNT_STREAMFLAGS_EVENTCALLBACK ? "EVENT" : "POLL")));
-
-    return paNoError;
-
-error:
-
-    return result;
-}
-
-// ------------------------------------------------------------------------------------------
-static PaError ActivateAudioClientInput(PaWasapiStream *stream)
+static PaError ActivateAudioClient(PaWasapiStream *stream, BOOL output)
 {
     HRESULT hr;
     PaError result;
     UINT32 maxBufferSize;
     PaTime bufferLatency;
-    const UINT32 framesPerBuffer = stream->in.params.frames_per_buffer;
+    PaWasapiSubStream *subStream = (output ? &stream->out : &stream->in);
 
     // Create Audio client
-    if (FAILED(hr = CreateAudioClient(stream, &stream->in, FALSE, &result)))
+    if (FAILED(hr = CreateAudioClient(stream, subStream, output, &result)))
     {
         LogPaError(result);
         goto error;
     }
-    LogWAVEFORMATEXTENSIBLE(&stream->in.wavex);
-
-    // Create volume mgr
-    stream->inVol = NULL;
-    /*hr = info->device->Activate(
-        __uuidof(IAudioEndpointVolume), CLSCTX_INPROC_SERVER, NULL,
-        (void**)&stream->inVol);
-    if (hr != S_OK)
-        return paInvalidDevice;*/
+    LogWAVEFORMATEXTENSIBLE(&subStream->wavexu.ext);
 
     // Get max possible buffer size to check if it is not less than that we request
-    if (FAILED(hr = IAudioClient_GetBufferSize(stream->in.clientParent, &maxBufferSize)))
+    if (FAILED(hr = IAudioClient_GetBufferSize(subStream->clientParent, &maxBufferSize)))
     {
         LogHostError(hr);
         LogPaError(result = paInvalidDevice);
@@ -3688,32 +3814,39 @@ static PaError ActivateAudioClientInput(PaWasapiStream *stream)
     }
 
     // Correct buffer to max size if it maxed out result of GetBufferSize
-    stream->in.bufferSize = maxBufferSize;
+    subStream->bufferSize = maxBufferSize;
 
-    // Get interface latency (actually unneeded as we calculate latency from the size
-    // of maxBufferSize).
-    if (FAILED(hr = IAudioClient_GetStreamLatency(stream->in.clientParent, &stream->in.deviceLatency)))
+    if (!output)
     {
-        LogHostError(hr);
-        LogPaError(result = paInvalidDevice);
-        goto error;
+        // Get interface latency (actually unneeded as we calculate latency from the size of maxBufferSize)
+        if (FAILED(hr = IAudioClient_GetStreamLatency(stream->in.clientParent, &stream->in.deviceLatency)))
+        {
+            LogHostError(hr);
+            LogPaError(result = paInvalidDevice);
+            goto error;
+        }
+        //stream->in.latencySeconds = nano100ToSeconds(stream->in.deviceLatency);
     }
-    //stream->in.latencySeconds = nano100ToSeconds(stream->in.deviceLatency);
 
     // Number of frames that are required at each period
-    stream->in.framesPerHostCallback = maxBufferSize;
+    subStream->framesPerHostCallback = maxBufferSize;
 
     // Calculate frames per single buffer, if buffers > 1 then always framesPerBuffer
-    stream->in.framesPerBuffer =
-        (stream->in.userBufferAndHostMatch ? stream->in.framesPerHostCallback : framesPerBuffer);
+    subStream->framesPerBuffer = (subStream->userBufferAndHostMatch ?
+        subStream->framesPerHostCallback : subStream->params.frames_per_buffer);
 
     // Calculate buffer latency
-    bufferLatency = (PaTime)maxBufferSize / stream->in.wavex.Format.nSamplesPerSec;
+    bufferLatency = (PaTime)maxBufferSize / subStream->wavexu.ext.Format.nSamplesPerSec;
 
     // Append buffer latency to interface latency in shared mode (see GetStreamLatency notes)
-    stream->in.latencySeconds = bufferLatency;
+    subStream->latencySeconds = bufferLatency;
 
-    PRINT(("WASAPI::OpenStream(input): framesPerUser[ %d ] framesPerHost[ %d ] latency[ %.02fms ] exclusive[ %s ] wow64_fix[ %s ] mode[ %s ]\n", (UINT32)framesPerBuffer, (UINT32)stream->in.framesPerHostCallback, (float)(stream->in.latencySeconds*1000.0f), (stream->in.shareMode == AUDCLNT_SHAREMODE_EXCLUSIVE ? "YES" : "NO"), (stream->in.params.wow64_workaround ? "YES" : "NO"), (stream->in.streamFlags & AUDCLNT_STREAMFLAGS_EVENTCALLBACK ? "EVENT" : "POLL")));
+    PRINT(("WASAPI::OpenStream(%s): framesPerUser[ %d ] framesPerHost[ %d ] latency[ %.02fms ] exclusive[ %s ] wow64_fix[ %s ] mode[ %s ]\n",
+        (output ? "output" : "input"),
+        subStream->params.frames_per_buffer, subStream->framesPerHostCallback, (subStream->latencySeconds * 1000),
+        (subStream->shareMode == AUDCLNT_SHAREMODE_EXCLUSIVE ? "YES" : "NO"),
+        (subStream->params.wow64_workaround ? "YES" : "NO"),
+        (subStream->streamFlags & AUDCLNT_STREAMFLAGS_EVENTCALLBACK ? "EVENT" : "POLL")));
 
     return paNoError;
 
@@ -3765,6 +3898,10 @@ static PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
         goto error;
     }
 
+    // Set stream state
+    stream->isActive  = FALSE;
+    stream->isStopped = TRUE;
+
     // Default thread priority is Audio: for exclusive mode we will use Pro Audio.
     stream->nThreadPriority = eThreadPriorityAudio;
 
@@ -3850,6 +3987,11 @@ static PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
             ((inputStreamInfo != NULL) && (inputStreamInfo->flags & paWinWasapiAutoConvert)))
             stream->in.streamFlags |= (AUDCLNT_STREAMFLAGS_AUTOCONVERTPCM | AUDCLNT_STREAMFLAGS_SRC_DEFAULT_QUALITY);
 
+        // Loopback device is a real output device, so we open stream with AUDCLNT_STREAMFLAGS_LOOPBACK
+        // flag to make it work as input device
+        if (info->loopBack)
+            stream->in.streamFlags |= AUDCLNT_STREAMFLAGS_LOOPBACK;
+
         // Fill parameters for Audio Client creation
         stream->in.params.device_info       = info;
         stream->in.params.stream_params     = (*inputParameters);
@@ -3860,14 +4002,14 @@ static PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
         stream->in.params.wow64_workaround  = paWasapi->useWOW64Workaround;
 
         // Create and activate audio client
-        if ((result = ActivateAudioClientInput(stream)) != paNoError)
+        if ((result = ActivateAudioClient(stream, FALSE)) != paNoError)
         {
             LogPaError(result);
             goto error;
         }
 
         // Get closest format
-        hostInputSampleFormat = PaUtil_SelectClosestAvailableFormat(WaveToPaFormat(&stream->in.wavex), inputSampleFormat);
+        hostInputSampleFormat = PaUtil_SelectClosestAvailableFormat(WaveToPaFormat(&stream->in.wavexu.ext), inputSampleFormat);
 
         // Set user-side custom host processor
         if ((inputStreamInfo != NULL) &&
@@ -3890,7 +4032,7 @@ static PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
         if (stream->in.params.blocking == TRUE)
         {
             UINT32 bufferFrames = ALIGN_NEXT_POW2((stream->in.framesPerHostCallback / WASAPI_PACKETS_PER_INPUT_BUFFER) * 2);
-            UINT32 frameSize    = stream->in.wavex.Format.nBlockAlign;
+            UINT32 frameSize    = stream->in.wavexu.ext.Format.nBlockAlign;
 
             // buffer
             if ((stream->in.tailBuffer = PaUtil_AllocateMemory(sizeof(PaUtilRingBuffer))) == NULL)
@@ -3994,14 +4136,14 @@ static PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
         stream->out.params.wow64_workaround  = paWasapi->useWOW64Workaround;
 
         // Create and activate audio client
-        if ((result = ActivateAudioClientOutput(stream)) != paNoError)
+        if ((result = ActivateAudioClient(stream, TRUE)) != paNoError)
         {
             LogPaError(result);
             goto error;
         }
 
         // Get closest format
-        hostOutputSampleFormat = PaUtil_SelectClosestAvailableFormat(WaveToPaFormat(&stream->out.wavex), outputSampleFormat);
+        hostOutputSampleFormat = PaUtil_SelectClosestAvailableFormat(WaveToPaFormat(&stream->out.wavexu.ext), outputSampleFormat);
 
         // Set user-side custom host processor
         if ((outputStreamInfo != NULL) &&
@@ -4049,14 +4191,14 @@ static PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
     // Initialize stream representation
     if (streamCallback)
     {
-        stream->bBlocking = FALSE;
+        stream->isBlocking = FALSE;
         PaUtil_InitializeStreamRepresentation(&stream->streamRepresentation,
                                               &paWasapi->callbackStreamInterface,
                                               streamCallback, userData);
     }
     else
     {
-        stream->bBlocking = TRUE;
+        stream->isBlocking = TRUE;
         PaUtil_InitializeStreamRepresentation(&stream->streamRepresentation,
                                               &paWasapi->blockingStreamInterface,
                                               streamCallback, userData);
@@ -4171,8 +4313,6 @@ static PaError CloseStream( PaStream* s )
     SAFE_RELEASE(stream->renderClientParent);
     SAFE_RELEASE(stream->out.clientParent);
     SAFE_RELEASE(stream->in.clientParent);
-    SAFE_RELEASE(stream->inVol);
-    SAFE_RELEASE(stream->outVol);
 
     CloseHandle(stream->event[S_INPUT]);
     CloseHandle(stream->event[S_OUTPUT]);
@@ -4413,8 +4553,11 @@ static PaError StartStream( PaStream *s )
         goto start_error;
     }
 
-    // Create thread
-    if (!stream->bBlocking)
+    // Set stream state
+    stream->isActive  = TRUE;
+    stream->isStopped = FALSE;
+
+    if (!stream->isBlocking)
     {
         // Create thread events
         stream->hThreadStart = CreateEvent(NULL, TRUE, FALSE, NULL);
@@ -4504,14 +4647,11 @@ static PaError StartStream( PaStream *s )
             }
         }
 
-        // Set parent to working pointers to use shared functions.
+        // Set parent to working pointers to use shared functions
         stream->captureClient  = stream->captureClientParent;
         stream->renderClient   = stream->renderClientParent;
         stream->in.clientProc  = stream->in.clientParent;
         stream->out.clientProc = stream->out.clientParent;
-
-        // Signal: stream running.
-        stream->running = TRUE;
     }
 
     return result;
@@ -4530,10 +4670,10 @@ start_error:
 }
 
 // ------------------------------------------------------------------------------------------
-void _StreamFinish(PaWasapiStream *stream)
+static void StopStreamByUser(PaWasapiStream *stream)
 {
     // Issue command to thread to stop processing and wait for thread exit
-    if (!stream->bBlocking)
+    if (!stream->isBlocking)
     {
         SignalObjectAndWait(stream->hCloseRequest, stream->hThreadExit, INFINITE, FALSE);
     }
@@ -4553,7 +4693,9 @@ void _StreamFinish(PaWasapiStream *stream)
     // Cleanup handles
     _StreamCleanup(stream);
 
-    stream->running = FALSE;
+    // Set stream state
+    stream->isActive  = FALSE;
+    stream->isStopped = TRUE;
 }
 
 // ------------------------------------------------------------------------------------------
@@ -4572,7 +4714,7 @@ void _StreamCleanup(PaWasapiStream *stream)
 static PaError StopStream( PaStream *s )
 {
     // Finish stream
-    _StreamFinish((PaWasapiStream *)s);
+    StopStreamByUser((PaWasapiStream *)s);
     return paNoError;
 }
 
@@ -4580,20 +4722,20 @@ static PaError StopStream( PaStream *s )
 static PaError AbortStream( PaStream *s )
 {
     // Finish stream
-    _StreamFinish((PaWasapiStream *)s);
+    StopStreamByUser((PaWasapiStream *)s);
     return paNoError;
 }
 
 // ------------------------------------------------------------------------------------------
 static PaError IsStreamStopped( PaStream *s )
 {
-    return !((PaWasapiStream *)s)->running;
+    return ((PaWasapiStream *)s)->isStopped;
 }
 
 // ------------------------------------------------------------------------------------------
 static PaError IsStreamActive( PaStream *s )
 {
-    return ((PaWasapiStream *)s)->running;
+    return ((PaWasapiStream *)s)->isActive;
 }
 
 // ------------------------------------------------------------------------------------------
@@ -4613,6 +4755,18 @@ static double GetStreamCpuLoad( PaStream* s )
     return PaUtil_GetCpuLoad(&((PaWasapiStream *)s)->cpuLoadMeasurer);
 }
 
+// ------------------------------------------------------------------------------------------
+static inline BOOL CheckForStopOrWait(PaWasapiStream *stream, UINT32 delay)
+{
+    return (WaitForSingleObject(stream->hCloseRequest, delay) != WAIT_TIMEOUT);
+}
+
+// ------------------------------------------------------------------------------------------
+static inline BOOL CheckForStop(PaWasapiStream *stream)
+{
+    return CheckForStopOrWait(stream, 0);
+}
+
 // ------------------------------------------------------------------------------------------
 static PaError ReadStream( PaStream* s, void *_buffer, unsigned long frames )
 {
@@ -4627,7 +4781,7 @@ static PaError ReadStream( PaStream* s, void *_buffer, unsigned long frames )
     ThreadIdleScheduler sched;
 
     // validate
-    if (!stream->running)
+    if (!stream->isActive)
         return paStreamIsStopped;
     if (stream->captureClient == NULL)
         return paBadStreamPtr;
@@ -4699,7 +4853,7 @@ static PaError ReadStream( PaStream* s, void *_buffer, unsigned long frames )
     while (frames != 0)
     {
         // Check if blocking call must be interrupted
-        if (WaitForSingleObject(stream->hCloseRequest, sleep) != WAIT_TIMEOUT)
+        if (CheckForStopOrWait(stream, sleep))
             break;
 
         // Get available frames (must be finding out available frames before call to IAudioCaptureClient_GetBuffer
@@ -4720,7 +4874,7 @@ static PaError ReadStream( PaStream* s, void *_buffer, unsigned long frames )
             {
                 UINT32 sleep_frames = (frames < stream->in.framesPerHostCallback ? frames : stream->in.framesPerHostCallback);
 
-                sleep  = GetFramesSleepTime(sleep_frames, stream->in.wavex.Format.nSamplesPerSec);
+                sleep  = GetFramesSleepTime(sleep_frames, stream->in.wavexu.ext.Format.nSamplesPerSec);
                 sleep /= 4; // wait only for 1/4 of the buffer
 
                 // WASAPI input provides packets, thus expiring packet will result in bad audio
@@ -4770,7 +4924,7 @@ static PaError ReadStream( PaStream* s, void *_buffer, unsigned long frames )
         // Save tail into buffer
         if ((frames == 0) && (available > processed))
         {
-            UINT32 bytes_processed = processed * stream->in.wavex.Format.nBlockAlign;
+            UINT32 bytes_processed = processed * stream->in.wavexu.ext.Format.nBlockAlign;
             UINT32 frames_to_save  = available - processed;
 
             PaUtil_WriteRingBuffer(stream->in.tailBuffer, wasapi_buffer + bytes_processed, frames_to_save);
@@ -4797,7 +4951,6 @@ static PaError WriteStream( PaStream* s, const void *_buffer, unsigned long fram
 {
     PaWasapiStream *stream = (PaWasapiStream*)s;
 
-    //UINT32 frames;
     const BYTE *user_buffer = (const BYTE *)_buffer;
     BYTE *wasapi_buffer;
     HRESULT hr = S_OK;
@@ -4806,7 +4959,7 @@ static PaError WriteStream( PaStream* s, const void *_buffer, unsigned long fram
     ThreadIdleScheduler sched;
 
     // validate
-    if (!stream->running)
+    if (!stream->isActive)
         return paStreamIsStopped;
     if (stream->renderClient == NULL)
         return paBadStreamPtr;
@@ -4834,7 +4987,7 @@ static PaError WriteStream( PaStream* s, const void *_buffer, unsigned long fram
     while (frames != 0)
     {
         // Check if blocking call must be interrupted
-        if (WaitForSingleObject(stream->hCloseRequest, sleep) != WAIT_TIMEOUT)
+        if (CheckForStopOrWait(stream, sleep))
             break;
 
         // Get frames available
@@ -4849,7 +5002,7 @@ static PaError WriteStream( PaStream* s, const void *_buffer, unsigned long fram
         {
             UINT32 sleep_frames = (frames < stream->out.framesPerHostCallback ? frames : stream->out.framesPerHostCallback);
 
-            sleep  = GetFramesSleepTime(sleep_frames, stream->out.wavex.Format.nSamplesPerSec);
+            sleep  = GetFramesSleepTime(sleep_frames, stream->out.wavexu.ext.Format.nSamplesPerSec);
             sleep /= 2; // wait only for half of the buffer
 
             // Avoid busy waiting, schedule next 1 millesecond wait
@@ -4920,7 +5073,7 @@ static signed long GetStreamReadAvailable( PaStream* s )
     UINT32  available = 0;
 
     // validate
-    if (!stream->running)
+    if (!stream->isActive)
         return paStreamIsStopped;
     if (stream->captureClient == NULL)
         return paBadStreamPtr;
@@ -4946,7 +5099,7 @@ static signed long GetStreamWriteAvailable( PaStream* s )
     UINT32  available = 0;
 
     // validate
-    if (!stream->running)
+    if (!stream->isActive)
         return paStreamIsStopped;
     if (stream->renderClient == NULL)
         return paBadStreamPtr;
@@ -4987,7 +5140,7 @@ static void WaspiHostProcessingLoop( void *inputBuffer,  long inputFrames,
     {
         PaTime pending_time;
         if ((hr = IAudioClient_GetCurrentPadding(stream->in.clientProc, &pending)) == S_OK)
-            pending_time = (PaTime)pending / (PaTime)stream->in.wavex.Format.nSamplesPerSec;
+            pending_time = (PaTime)pending / (PaTime)stream->in.wavexu.ext.Format.nSamplesPerSec;
         else
             pending_time = (PaTime)stream->in.latencySeconds;
 
@@ -4998,7 +5151,7 @@ static void WaspiHostProcessingLoop( void *inputBuffer,  long inputFrames,
     {
         PaTime pending_time;
         if ((hr = IAudioClient_GetCurrentPadding(stream->out.clientProc, &pending)) == S_OK)
-            pending_time = (PaTime)pending / (PaTime)stream->out.wavex.Format.nSamplesPerSec;
+            pending_time = (PaTime)pending / (PaTime)stream->out.wavexu.ext.Format.nSamplesPerSec;
         else
             pending_time = (PaTime)stream->out.latencySeconds;
 
@@ -5052,19 +5205,9 @@ static void WaspiHostProcessingLoop( void *inputBuffer,  long inputFrames,
 
     PaUtil_EndCpuLoadMeasurement( &stream->cpuLoadMeasurer, framesProcessed );
 
-    if (callbackResult == paContinue)
-    {
-        /* nothing special to do */
-    }
-    else
-    if (callbackResult == paAbort)
-    {
-        // stop stream
-        SetEvent(stream->hCloseRequest);
-    }
-    else
+    if (callbackResult != paContinue)
     {
-        // stop stream
+        // schedule stream stop
         SetEvent(stream->hCloseRequest);
     }
 }
@@ -5564,21 +5707,19 @@ PaError PaWasapi_SetStreamStateHandler( PaStream *pStream, PaWasapiStreamStateCa
 HRESULT _PollGetOutputFramesAvailable(PaWasapiStream *stream, UINT32 *available)
 {
     HRESULT hr;
-    UINT32 frames  = stream->out.framesPerHostCallback,
-           padding = 0;
+    UINT32 frames = stream->out.framesPerHostCallback,
+           padding;
 
     (*available) = 0;
 
-    // get read position
+    // Get read position
     if ((hr = IAudioClient_GetCurrentPadding(stream->out.clientProc, &padding)) != S_OK)
         return LogHostError(hr);
 
-    // get available
-    frames -= padding;
+    // Set available frame count
+    (*available) = frames - padding;
 
-    // set
-    (*available) = frames;
-    return hr;
+    return S_OK;
 }
 
 // ------------------------------------------------------------------------------------------
@@ -5592,14 +5733,14 @@ HRESULT _PollGetInputFramesAvailable(PaWasapiStream *stream, UINT32 *available)
     if ((hr = IAudioClient_GetCurrentPadding(stream->in.clientProc, available)) != S_OK)
         return LogHostError(hr);
 
-    return hr;
+    return S_OK;
 }
 
 // ------------------------------------------------------------------------------------------
 static HRESULT ProcessOutputBuffer(PaWasapiStream *stream, PaWasapiHostProcessor *processor, UINT32 frames)
 {
     HRESULT hr;
-    BYTE *data = NULL;
+    BYTE *data;
 
     // Get buffer
     if ((hr = IAudioRenderClient_GetBuffer(stream->renderClient, frames, &data)) != S_OK)
@@ -5644,23 +5785,19 @@ static HRESULT ProcessOutputBuffer(PaWasapiStream *stream, PaWasapiHostProcessor
     // Process data
     if (stream->out.monoMixer != NULL)
     {
-        // expand buffer
-        UINT32 mono_frames_size = frames * (stream->out.wavex.Format.wBitsPerSample / 8);
-        if (mono_frames_size > stream->out.monoBufferSize)
+        // Expand buffer
+        UINT32 monoFrames = frames * (stream->out.wavexu.ext.Format.wBitsPerSample / 8);
+        if (monoFrames > stream->out.monoBufferSize)
         {
-            stream->out.monoBuffer = PaWasapi_ReallocateMemory(stream->out.monoBuffer, (stream->out.monoBufferSize = mono_frames_size));
+            stream->out.monoBuffer = PaWasapi_ReallocateMemory(stream->out.monoBuffer, (stream->out.monoBufferSize = monoFrames));
             if (stream->out.monoBuffer == NULL)
-            {
-                hr = E_OUTOFMEMORY;
-                LogHostError(hr);
-                return hr;
-            }
+                return LogHostError(hr = E_OUTOFMEMORY);
         }
 
-        // process
+        // Process
         processor[S_OUTPUT].processor(NULL, 0, (BYTE *)stream->out.monoBuffer, frames, processor[S_OUTPUT].userData);
 
-        // mix 1 to 2 channels
+        // Mix 1 to 2 channels
         stream->out.monoMixer(data, stream->out.monoBuffer, frames);
     }
     else
@@ -5670,27 +5807,22 @@ static HRESULT ProcessOutputBuffer(PaWasapiStream *stream, PaWasapiHostProcessor
 
     // Release buffer
     if ((hr = IAudioRenderClient_ReleaseBuffer(stream->renderClient, frames, 0)) != S_OK)
-        LogHostError(hr);
+        return LogHostError(hr);
 
-    return hr;
+    return S_OK;
 }
 
 // ------------------------------------------------------------------------------------------
 static HRESULT ProcessInputBuffer(PaWasapiStream *stream, PaWasapiHostProcessor *processor)
 {
-    HRESULT hr = S_OK;
+    HRESULT hr;
     UINT32 frames;
-    BYTE *data = NULL;
-    DWORD flags = 0;
+    BYTE *data;
+    DWORD flags;
 
-    for (;;)
+    while (!CheckForStop(stream))
     {
-        // Check if blocking call must be interrupted
-        if (WaitForSingleObject(stream->hCloseRequest, 0) != WAIT_TIMEOUT)
-            break;
-
         // Find out if any frames available
-        frames = 0;
         if ((hr = _PollGetInputFramesAvailable(stream, &frames)) != S_OK)
             return hr;
 
@@ -5708,7 +5840,6 @@ static HRESULT ProcessInputBuffer(PaWasapiStream *stream, PaWasapiHostProcessor
             }
 
             return LogHostError(hr);
-            break;
         }
 
         // Detect silence
@@ -5719,16 +5850,12 @@ static HRESULT ProcessInputBuffer(PaWasapiStream *stream, PaWasapiHostProcessor
         if (stream->in.monoMixer != NULL)
         {
             // expand buffer
-            UINT32 mono_frames_size = frames * (stream->in.wavex.Format.wBitsPerSample / 8);
-            if (mono_frames_size > stream->in.monoBufferSize)
+            UINT32 monoFrames = frames * (stream->in.wavexu.ext.Format.wBitsPerSample / 8);
+            if (monoFrames > stream->in.monoBufferSize)
             {
-                stream->in.monoBuffer = PaWasapi_ReallocateMemory(stream->in.monoBuffer, (stream->in.monoBufferSize = mono_frames_size));
+                stream->in.monoBuffer = PaWasapi_ReallocateMemory(stream->in.monoBuffer, (stream->in.monoBufferSize = monoFrames));
                 if (stream->in.monoBuffer == NULL)
-                {
-                    hr = E_OUTOFMEMORY;
-                    LogHostError(hr);
-                    return hr;
-                }
+                    return LogHostError(hr = E_OUTOFMEMORY);
             }
 
             // mix 1 to 2 channels
@@ -5745,18 +5872,16 @@ static HRESULT ProcessInputBuffer(PaWasapiStream *stream, PaWasapiHostProcessor
         // Release buffer
         if ((hr = IAudioCaptureClient_ReleaseBuffer(stream->captureClient, frames)) != S_OK)
             return LogHostError(hr);
-
-        //break;
     }
 
-    return hr;
+    return S_OK;
 }
 
 // ------------------------------------------------------------------------------------------
 void _StreamOnStop(PaWasapiStream *stream)
 {
     // Stop INPUT/OUTPUT clients
-    if (!stream->bBlocking)
+    if (!stream->isBlocking)
     {
         if (stream->in.clientProc != NULL)
             IAudioClient_Stop(stream->in.clientProc);
@@ -5840,7 +5965,7 @@ PA_THREAD_FUNC ProcThreadEvent(void *param)
     BOOL threadComInitialized = FALSE;
     SystemTimer timer;
 
-    // Notify: state
+    // Notify: WASAPI-specific stream state
     NotifyStateChanged(stream, paWasapiStreamStateThreadPrepare, ERROR_SUCCESS);
 
     // Prepare COM pointers
@@ -5883,8 +6008,8 @@ PA_THREAD_FUNC ProcThreadEvent(void *param)
         goto thread_error;
     }
 
-    // Signal: stream running
-    stream->running = TRUE;
+    // Signal: stream active (reconfirm)
+    stream->isActive = TRUE;
 
     // Notify: thread started
     SetEvent(stream->hThreadStart);
@@ -5939,7 +6064,7 @@ PA_THREAD_FUNC ProcThreadEvent(void *param)
 
     }
 
-    // Notify: state
+    // Notify: WASAPI-specific stream state
     NotifyStateChanged(stream, paWasapiStreamStateThreadStart, ERROR_SUCCESS);
 
     // Processing Loop
@@ -5950,7 +6075,7 @@ PA_THREAD_FUNC ProcThreadEvent(void *param)
 
         // Check for close event (after wait for buffers to avoid any calls to user
         // callback when hCloseRequest was set)
-        if (WaitForSingleObject(stream->hCloseRequest, 0) != WAIT_TIMEOUT)
+        if (CheckForStop(stream))
             break;
 
         // Process S_INPUT/S_OUTPUT
@@ -6002,13 +6127,13 @@ thread_end:
     // Restore system timer granularity
     SystemTimer_RestoreGranularity(&timer);
 
-    // Notify: not running
-    stream->running = FALSE;
+    // Notify: stream inactive
+    stream->isActive = FALSE;
 
     // Notify: thread exited
     SetEvent(stream->hThreadExit);
 
-    // Notify: state
+    // Notify: WASAPI-specific stream state
     NotifyStateChanged(stream, paWasapiStreamStateThreadStop, hr);
 
     return 0;
@@ -6065,8 +6190,8 @@ static UINT32 ConfigureLoopSleepTimeAndScheduler(PaWasapiStream *stream, ThreadI
     }
 
     // Calculate timeout for the next polling attempt
-    sleepTimeIn  = GetFramesSleepTime(userFramesIn, stream->in.wavex.Format.nSamplesPerSec);
-    sleepTimeOut = GetFramesSleepTime(userFramesOut, stream->out.wavex.Format.nSamplesPerSec);
+    sleepTimeIn  = GetFramesSleepTime(userFramesIn, stream->in.wavexu.ext.Format.nSamplesPerSec);
+    sleepTimeOut = GetFramesSleepTime(userFramesOut, stream->out.wavexu.ext.Format.nSamplesPerSec);
 
     // WASAPI input packets tend to expire very easily, let's limit sleep time to 2 milliseconds
     // for all cases. Please propose better solution if any
@@ -6078,8 +6203,8 @@ static UINT32 ConfigureLoopSleepTimeAndScheduler(PaWasapiStream *stream, ThreadI
     // Make sure not 0, othervise use ThreadIdleScheduler to bounce between [0, 1] ms to avoid too busy loop
     if (sleepTime == 0)
     {
-        sleepTimeIn  = GetFramesSleepTimeMicroseconds(userFramesIn, stream->in.wavex.Format.nSamplesPerSec);
-        sleepTimeOut = GetFramesSleepTimeMicroseconds(userFramesOut, stream->out.wavex.Format.nSamplesPerSec);
+        sleepTimeIn  = GetFramesSleepTimeMicroseconds(userFramesIn, stream->in.wavexu.ext.Format.nSamplesPerSec);
+        sleepTimeOut = GetFramesSleepTimeMicroseconds(userFramesOut, stream->out.wavexu.ext.Format.nSamplesPerSec);
 
         sleepTime = GetSleepTime(stream, sleepTimeIn, sleepTimeOut, userFramesOut);
 
@@ -6143,7 +6268,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
     LONGLONG startWaitTime;
 #endif
 
-    // Notify: state
+    // Notify: WASAPI-specific stream state
     NotifyStateChanged(stream, paWasapiStreamStateThreadPrepare, ERROR_SUCCESS);
 
     // Prepare COM pointers
@@ -6165,8 +6290,8 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
     // Boost thread priority
     PaWasapi_ThreadPriorityBoost((void **)&stream->hAvTask, stream->nThreadPriority);
 
-    // Signal: stream running
-    stream->running = TRUE;
+    // Signal: stream active (reconfirm)
+    stream->isActive = TRUE;
 
     // Notify: thread started
     SetEvent(stream->hThreadStart);
@@ -6195,7 +6320,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
                 {
                     // It is important to preload whole host buffer to avoid underruns/glitches when stream is started,
                     // for more details see the discussion: https://github.com/PortAudio/portaudio/issues/303
-                    while (frames >= stream->out.framesPerBuffer)
+                    while ((frames >= stream->out.framesPerBuffer) && !CheckForStop(stream))
                     {
                         if ((hr = ProcessOutputBuffer(stream, processor, stream->out.framesPerBuffer)) != S_OK)
                         {
@@ -6238,7 +6363,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
         }
     }
 
-    // Notify: state
+    // Notify: WASAPI-specific stream state
     NotifyStateChanged(stream, paWasapiStreamStateThreadStart, ERROR_SUCCESS);
 
 #ifdef PA_WASAPI_LOG_TIME_SLOTS
@@ -6248,7 +6373,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
     if (!PA_WASAPI__IS_FULLDUPLEX(stream))
     {
         // Processing Loop
-        while (WaitForSingleObject(stream->hCloseRequest, nextSleepTime) == WAIT_TIMEOUT)
+        while (!CheckForStopOrWait(stream, nextSleepTime))
         {
             startTime = SystemTimer_GetTime(&timer);
 
@@ -6305,7 +6430,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
                             continue;
                         }
 
-                        while (framesAvail >= framesProc)
+                        do
                         {
                             if ((hr = ProcessOutputBuffer(stream, processor, framesProc)) != S_OK)
                             {
@@ -6315,6 +6440,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
 
                             framesAvail -= framesProc;
                         }
+                        while ((framesAvail >= framesProc) && !CheckForStop(stream));
                     }
                     else
                     if (framesAvail != 0)
@@ -6341,7 +6467,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
     else
     {
         // Processing Loop (full-duplex)
-        while (WaitForSingleObject(stream->hCloseRequest, nextSleepTime) == WAIT_TIMEOUT)
+        while (!CheckForStopOrWait(stream, nextSleepTime))
         {
             UINT32 i_frames = 0, i_processed = 0, o_frames = 0;
             BYTE *i_data = NULL, *o_data = NULL, *o_data_host = NULL;
@@ -6388,7 +6514,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
                         // convert output mono
                         if (stream->out.monoMixer)
                         {
-                            UINT32 mono_frames_size = o_processed * (stream->out.wavex.Format.wBitsPerSample / 8);
+                            UINT32 mono_frames_size = o_processed * (stream->out.wavexu.ext.Format.wBitsPerSample / 8);
                             // expand buffer
                             if (mono_frames_size > stream->out.monoBufferSize)
                             {
@@ -6412,7 +6538,7 @@ PA_THREAD_FUNC ProcThreadPoll(void *param)
                         // convert input mono
                         if (stream->in.monoMixer)
                         {
-                            UINT32 mono_frames_size = i_processed * (stream->in.wavex.Format.wBitsPerSample / 8);
+                            UINT32 mono_frames_size = i_processed * (stream->in.wavexu.ext.Format.wBitsPerSample / 8);
                             // expand buffer
                             if (mono_frames_size > stream->in.monoBufferSize)
                             {
@@ -6495,13 +6621,13 @@ thread_end:
     // Restore system timer granularity
     SystemTimer_RestoreGranularity(&timer);
 
-    // Notify: not running
-    stream->running = FALSE;
+    // Notify: state inactive
+    stream->isActive = FALSE;
 
     // Notify: thread exited
     SetEvent(stream->hThreadExit);
 
-    // Notify: state
+    // Notify: WASAPI-specific stream state
     NotifyStateChanged(stream, paWasapiStreamStateThreadStop, hr);
 
     return 0;
diff --git a/test/patest_callbackstop.c b/test/patest_callbackstop.c
index fba9dca..3f4152b 100644
--- a/test/patest_callbackstop.c
+++ b/test/patest_callbackstop.c
@@ -123,13 +123,36 @@ static void StreamFinished( void* userData )
     printf( "Stream Completed: %s\n", data->message );
 }
 
+/**
+ * @return paNoError if successful, otherwise a negative error.
+ */
+static PaError CheckActiveStopped(PaStream *stream,
+                                  PaError expectedActive,
+                                  PaError expectedStopped,
+                                  int lineNumber)
+{
+    PaError testResult = paNoError;
+    PaError actualActive = Pa_IsStreamActive(stream);
+    if (expectedActive != actualActive) {
+        printf("ERROR at line %d - active is %d, expected %d\n",
+               lineNumber, actualActive, expectedActive);
+        testResult = (actualActive < 0) ? actualActive : paInternalError;
+    }
+    PaError actualStopped = Pa_IsStreamStopped(stream);
+    if (expectedStopped != actualStopped) {
+        printf("ERROR at line %d - stopped is %d, expected %d\n",
+               lineNumber, actualStopped, expectedStopped);
+        testResult = (actualStopped < 0) ? actualStopped : paInternalError;
+    }
+    return testResult;
+}
 
 /*----------------------------------------------------------------------------*/
 int main(void);
 int main(void)
 {
     PaStreamParameters outputParameters;
-    PaStream *stream;
+    PaStream *stream = NULL;
     PaError err;
     TestData data;
     int i, j;
@@ -147,7 +170,7 @@ int main(void)
     err = Pa_Initialize();
     if( err != paNoError ) goto error;
 
-    outputParameters.device                    = Pa_GetDefaultOutputDevice();
+    outputParameters.device = Pa_GetDefaultOutputDevice();
     if (outputParameters.device == paNoDevice) {
         fprintf(stderr,"Error: No default output device.\n");
         goto error;
@@ -172,6 +195,7 @@ int main(void)
     err = Pa_SetStreamFinishedCallback( stream, &StreamFinished );
     if( err != paNoError ) goto error;
 
+
     printf("Repeating test %d times.\n", NUM_LOOPS );
 
     for( i=0; i < NUM_LOOPS; ++i )
@@ -182,9 +206,13 @@ int main(void)
         data.callbackInvokedAfterReturningPaComplete = 0;
         sprintf( data.message, "Loop: %d", i );
 
+        if( (err = CheckActiveStopped(stream, 0, 1, __LINE__)) < 0 ) goto error;
+
         err = Pa_StartStream( stream );
         if( err != paNoError ) goto error;
 
+        if( (err = CheckActiveStopped(stream, 1, 0, __LINE__)) < 0 ) goto error;
+
         printf("Play for %d seconds.\n", NUM_SECONDS );
 
         /* wait for the callback to complete generating NUM_SECONDS of tone */
@@ -228,10 +256,14 @@ int main(void)
             printf( "TEST FAILED: Callback was invoked after returning paComplete.\n" );
         }
 
+        /* Stream should not be "stopped" until Pa_StopStream() called. */
+        if( (err = CheckActiveStopped(stream, 0, 0, __LINE__)) < 0 ) goto error;
 
         err = Pa_StopStream( stream );
         if( err != paNoError ) goto error;
 
+        if( (err = CheckActiveStopped(stream, 0, 1, __LINE__)) < 0 ) goto error;
+
         printf( "sleeping for 1 second...\n" );
         Pa_Sleep( 1000 );
     }
@@ -244,6 +276,10 @@ int main(void)
 
     return err;
 error:
+    if( stream != NULL )
+    {
+        Pa_CloseStream( stream );
+    }
     Pa_Terminate();
     fprintf( stderr, "An error occurred while using the portaudio stream\n" );
     fprintf( stderr, "Error number: %d\n", err );
